Add wxLunaSVG rendering enginge library

Randalphwa opened the pull request:
This _draft_ PR is related to issue #24216 -- it is designed to provide the option of replacing the NanoSVG rendering engine with the somewhat faster and definitely more accurate wxLunaSVG rendering engine.

By default, the wxWidgets build will not change, and NanoSVG will still be used. To replace it with wxLunaSVG when building with CMake, you need to do three things:

1) Run git submodule update --init 3rdparty/lunasvg
2) Set the wxBUILD_CXX_STANDARD to 17 or higher
3) Change the wxUSE_LUNASVG option to ON.

While I have updated the necessary bakefiles, it's been a _long_ time since I have worked with makefiles or vs project files, so I still need to look further into this to see how to get it to build under those build systems. Hence the PR being a draft...

Randalphwa committed:
Add updated lunasvg submodule

Randalphwa committed:
Add wxUSE_LUNASVG

Default to 0 because of the requirement for a C++17 compiler

Randalphwa committed:
Add wxWidgets and lunasvg code

Randalphwa committed:
Add configuration and docs

Randalphwa committed:
Add lunasvg submodule

Randalphwa committed:
Support USE_LUNASVG for building wxlunasvg library

Randalphwa committed:
Add wxlunasvg to bakefiles

PBfordev commented:
IMO, the code maintainability could be improved by removing the duplicate code existing for the two implementations.

I would factor out and share the common code:
1. Common wx include headers
2. `wxBitmapBundleImplSVG` declaration and all its methods implementations (e.g., inlined) besides the ctor and `DoRasterize()` as well as `m_sizeDef` and `m_cachedBitmap` variables
3. Most `wxBitmapBundle::From*()` method definitions

If you want, I could give it a try, provided CMake build on MSW works.

Randalphwa committed:
Generate FATAL_ERROR if building without C++17

Thanks to MaartenBent for the suggestion!

Randalphwa committed:
Move common headers outside of luna conditionals

Thanks to PBfordev for code review suggestion!

Randalphwa commented:
> PBfordev

Thanks for the suggestions!

I've moved both common headers and code to remove duplicates. I want to verify the changes with some apps I have that use both rendering engines before I check it in, probably not until this weekend.

Randalphwa commented:
> > I'm probably being dense here, but what is the purpose of:
> 
> This is what links `wxlunasvg` to `wxcore`. Together with this modification in `wxcore`:
> 
> ```diff
> -foreach(lib JPEG PNG TIFF NANOSVG WebP)
> +foreach(lib JPEG PNG TIFF NANOSVG WebP LUNASVG)
> ```
> 
> it checks if `LUNASVG_LIBRARIES ` and/or `LUNASVG_INCLUDE_DIRS` exists, and if they exist it will add the dependency to `wxcore`.

I'm inclined to add something along the lines of:

```cmake
if(NOT wxUSE_LUNASVG STREQUAL "OFF" AND (CMAKE_CXX_STANDARD GREATER_EQUAL 17 OR wxHAVE_CXX17))
    set(HAVE_LUNASVG "WXLUNASVG")
endif()

wx_add_library(wxcore ${CORE_SRC})
foreach(lib JPEG PNG TIFF NANOSVG WebP ${HAVE_LUNASVG})
```

That way it's clear that the library isn't always available.

MaartenBent commented:
Any of the third party libraries might be unavailable, that's why there are checks inside the `foreach` loop. I prefer to handle all of the third party libraries the same instead of having this special check for lunasvg.
And keep all the lunasvg code inside the lunasvg cmake file.

vadz reviewed (commented):
Thanks for your work on this!

I think we need at least 1, and preferably 2 (1 under Unix and 1 under Windows) CI jobs testing this to detect any breakage to non-default build. Could you please modify `.github/workflows/ci_cmake.yml` and `.github/workflows/ci.yml` to add `use_lunasvg` parameter to the builds matrix there and use it if it is set to true? TIA!

Randalphwa commented on .cirrus.yml:
```diff
@@ -56,7 +56,7 @@ task:
   # use system libraries instead of the other ones.
   update_submodules_script: |
-    git submodule update --init 3rdparty/catch 3rdparty/nanosvg src/stc/scintilla src/stc/lexilla
+    git submodule update --init 3rdparty/catch 3rdparty/nanosvg 3rdparty/lunasvg src/stc/scintilla src/stc/lexilla
```
I'll add the build CI jobs as requested, and modify when we pull in either the nanosvg or lunasvg modules based on the build. I haven't had a chance to look yet, but is there already a CI for testing against the C++17 compiler since we do have some conditionalized code in wxString that uses that? If so, would it make sense to add the wxlunasvg portion to that CI (or CIs) rather than creating separate CI jobs?

Randalphwa committed:
Add wxlunasvg.lib to core if available

Per MaartenBent's suggestion, this eliminates the need to specify the lunasvg library
in any app that enables wxUSE_LUNASVG.

Randalphwa committed:
Revert .circleci/config.yml, .cirrus.yml and src/zlib to state before de4ab3628 (keep other changes)

Randalphwa committed:
Refactor to minimize duplication

Place common headers outside of the wxUSE_LUNASVG conditional, add a common
LoadSVGFile(const wxString& pat) function.

Randalphwa committed:
Move LoadSVGFile() out of anonymous namespace

Randalphwa committed:
Add wxUSE_LUNASVG 1 to C++20 msw build

Since building wxlunasvg requires C++17 as a minimum, it makes sense to also build
wxlunasvg in the C++20 msw build to verify that it does in fact build.

Randalphwa commented on src/zlib:
I reverted the changes to the two CI files as well as this one.

Randalphwa commented:
> I think we need at least 1, and preferably 2 (1 under Unix and 1 under Windows) CI jobs testing this to detect any breakage to non-default build. Could you please modify `.github/workflows/ci_cmake.yml` and `.github/workflows/ci.yml` to add `use_lunasvg` parameter to the builds matrix there and use it if it is set to true? TIA!

I added #define wxUSE_LUNASVG to the msw build that uses c++20 since that way I knew it had the c++17 minimum compiler to build wxlunasvg. The `Ubuntu 24.04 wxGTK UBSAN` matrix also uses c++20, so if it makes sense to shoehorn the wxlunasvg builds into the cxx20 builds, I'll look into adding it to that matrix as well.

MaartenBent commented on build/cmake/lib/core/CMakeLists.txt:
```diff
@@ -60,7 +60,7 @@ elseif(WXX11)
 
 wx_add_library(wxcore ${CORE_SRC})
-foreach(lib JPEG PNG TIFF NANOSVG WebP)
+foreach(lib JPEG PNG TIFF NANOSVG WebP wxlunasvg)
```
Linking still fails for me. Because this should be `LUNASVG`, it needs to match the prefix of the `LUNASVG_LIBRARIES` and `LUNASVG_INCLUDE_DIRS` variables.
Do you use monolithic build if it succeeds for you?

```
16>wxmsw33ud_core.lib(bmpsvg.obj) : error LNK2019: unresolved external symbol "public: __cdecl wxlunasvg::Bitmap::~Bitmap(void)" (??1Bitmap@wxlunasvg@@QEAA@XZ) referenced in function "private: class wxBitmap __cdecl wxBitmapBundleImplSVG::DoRasterize(class wxSize const &)" (?DoRasterize@wxBitmapBundleImplSVG@@AEAA?AVwxBitmap@@AEBVwxSize@@@Z)
16>wxmsw33ud_core.lib(bmpsvg.obj) : error LNK2019: unresolved external symbol "public: unsigned char * __cdecl wxlunasvg::Bitmap::data(void)const " (?data@Bitmap@wxlunasvg@@QEBAPEAEXZ) referenced in function "private: class wxBitmap __cdecl wxBitmapBundleImplSVG::DoRasterize(class wxSize const &)" (?DoRasterize@wxBitmapBundleImplSVG@@AEAA?AVwxBitmap@@AEBVwxSize@@@Z)
16>wxmsw33ud_core.lib(bmpsvg.obj) : error LNK2019: unresolved external symbol "public: int __cdecl wxlunasvg::Bitmap::width(void)const " (?width@Bitmap@wxlunasvg@@QEBAHXZ) referenced in function "private: class wxBitmap __cdecl wxBitmapBundleImplSVG::DoRasterize(class wxSize const &)" (?DoRasterize@wxBitmapBundleImplSVG@@AEAA?AVwxBitmap@@AEBVwxSize@@@Z)
16>wxmsw33ud_core.lib(bmpsvg.obj) : error LNK2019: unresolved external symbol "public: int __cdecl wxlunasvg::Bitmap::height(void)const " (?height@Bitmap@wxlunasvg@@QEBAHXZ) referenced in function "private: class wxBitmap __cdecl wxBitmapBundleImplSVG::DoRasterize(class wxSize const &)" (?DoRasterize@wxBitmapBundleImplSVG@@AEAA?AVwxBitmap@@AEBVwxSize@@@Z)
16>wxmsw33ud_core.lib(bmpsvg.obj) : error LNK2019: unresolved external symbol "public: int __cdecl wxlunasvg::Bitmap::stride(void)const " (?stride@Bitmap@wxlunasvg@@QEBAHXZ) referenced in function "private: class wxBitmap __cdecl wxBitmapBundleImplSVG::DoRasterize(class wxSize const &)" (?DoRasterize@wxBitmapBundleImplSVG@@AEAA?AVwxBitmap@@AEBVwxSize@@@Z)
16>wxmsw33ud_core.lib(bmpsvg.obj) : error LNK2019: unresolved external symbol "public: static class std::unique_ptr<class wxlunasvg::Document,struct std::default_delete<class wxlunasvg::Document> > __cdecl wxlunasvg::Document::loadFromData(char const *)" (?loadFromData@Document@wxlunasvg@@SA?AV?$unique_ptr@VDocument@wxlunasvg@@U?$default_delete@VDocument@wxlunasvg@@@std@@@std@@PEBD@Z) referenced in function "public: __cdecl wxBitmapBundleImplSVG::wxBitmapBundleImplSVG(char const *,class wxSize const &)" (??0wxBitmapBundleImplSVG@@QEAA@PEBDAEBVwxSize@@@Z)
16>wxmsw33ud_core.lib(bmpsvg.obj) : error LNK2019: unresolved external symbol "public: static class std::unique_ptr<class wxlunasvg::Document,struct std::default_delete<class wxlunasvg::Document> > __cdecl wxlunasvg::Document::loadFromData(char const *,unsigned __int64)" (?loadFromData@Document@wxlunasvg@@SA?AV?$unique_ptr@VDocument@wxlunasvg@@U?$default_delete@VDocument@wxlunasvg@@@std@@@std@@PEBD_K@Z) referenced in function "public: __cdecl wxBitmapBundleImplSVG::wxBitmapBundleImplSVG(unsigned char const *,unsigned __int64,class wxSize const &)" (??0wxBitmapBundleImplSVG@@QEAA@PEBE_KAEBVwxSize@@@Z)
16>wxmsw33ud_core.lib(bmpsvg.obj) : error LNK2019: unresolved external symbol "public: class wxlunasvg::Bitmap __cdecl wxlunasvg::Document::renderToBitmap(int,int,unsigned int)const " (?renderToBitmap@Document@wxlunasvg@@QEBA?AVBitmap@2@HHI@Z) referenced in function "private: class wxBitmap __cdecl wxBitmapBundleImplSVG::DoRasterize(class wxSize const &)" (?DoRasterize@wxBitmapBundleImplSVG@@AEAA?AVwxBitmap@@AEBVwxSize@@@Z)
16>wxmsw33ud_core.lib(bmpsvg.obj) : error LNK2019: unresolved external symbol "public: __cdecl wxlunasvg::Document::~Document(void)" (??1Document@wxlunasvg@@QEAA@XZ) referenced in function "public: void * __cdecl wxlunasvg::Document::`scalar deleting destructor'(unsigned int)" (??_GDocument@wxlunasvg@@QEAAPEAXI@Z)
```

Randalphwa replied:
My bad -- I still had lunasvg as a required wxWidgets component in my test app. Corrected version is incoming...

Randalphwa committed:
Fix prefix for luna _LIBRRIES and _INCLUDE_DIRS

vadz commented:
> I can add documentation about how to enable wxlunasvg for these build systems, but I'm not sure where the best place to put that would be that someone could actually find. Any suggestions?

I'd say the best is to explain it in `interface/wx/bmpbndl.h`, i.e. add a section about configuring the build to use LunaSVG. This may not be the most logical place, but it should increase the probability of people finding this documentation.

vadz commented:
Oh, and I forgot to ask: would you like to preserve the existing commits or squash merge this entire PR as a single commit?

Randalphwa committed:
Add brief comment about how to enable LunaSVG library

Randalphwa committed:
Set wxUSE_LUNASVG  to 1 in cxx-20 Ubuntu build

The matrix.use_ubsan adds --with-cxx=20, so this commit also changes setup.h
to enable wxUSE_LUNASVG

MaartenBent commented on configure.ac:
```diff
@@ -541,6 +541,7 @@ WX_ARG_WITH(libnotify,     [  --with-libnotify        use libnotify for notifica
 WX_ARG_WITH(opengl,        [  --with-opengl           use OpenGL (or Mesa)], wxUSE_OPENGL)
 WX_ARG_WITH(xtest,         [  --with-xtest            use XTest extension], wxUSE_XTEST)
 WX_ARG_WITH(nanosvg,       [  --with-nanosvg          use NanoSVG for rasterizing SVG], wxUSE_NANOSVG)
+WX_ARG_WITH(lunasvg,       [  --with-lunasvg          use LunaSVG for rasterizing SVG], wxUSE_LUNASVG)
```
You've only modified this `configure.ac` file. You'll need to run `autogen.sh` to update the `configure` file.

MaartenBent replied:
I haven't used `configure` in a while, but I think it generates its own `setup.h` from `setup.h.in`.

I'll try it with lunasvg and report back.

Did you manually update the setup files or use `build/update-setup-h`?
(see `docs/contributing/how-to-add-new-wxUSE_XXX.md`)

MaartenBent commented:
> I don't see why this library would be any different than any other of the wxWidgets libraries that depend on a different wxWidgets library

Because wx libraries have no cyclic dependency, see for example the dependency picture in https://docs.wxwidgets.org/trunk/page_libs.html.

You mention png, and its exported symbols have indeed a `wx_` prefix. So it does make sense to do something similar to lunasvg. And I suppose changing namespace would do it. I can't think of any simpler, less intrusive way either.

Still, I'd like to see support for using system lunasvg as well, because it could be a nice upgrade from nanosvg. And we allow it for every third-party library.
You don't have to add it to this PR (maybe error if `sys` is used, or warning and force `builtin`).

But to prevent annoying cmake errors in the future, it would be beneficial if you could use `wx_add_thirdparty_library` (with default option `builtin`) instead of `wx_option`. Because the first one has allowed options `builtin`, `sys` and `off`, and the other has allowed options `on` and `off`.

asmwarrior commented:
Hello, I'd like to ask a question about how to modify the svg image at runtime. I mean the lunasvg library support modify or apply the styles, such as the colors can be changed, see here:

[How to tweak the svg image, and later update bundle bitmap · Issue #1 · PBfordev/wxtestsvg2](https://github.com/PBfordev/wxtestsvg2/issues/1)

So, the svg file is parsed once, and the styles can change many times. see: 

https://github.com/sammycage/lunasvg?tab=readme-ov-file#dynamic-styling

So, is there any way to support this level of operation? Thanks.

Randalphwa commented:
I added it as a separate issue since while lunasvg would support the style change, it would need to work with wxWidgets to have controls process it as per @PBfordev comment to that effect.

Randalphwa committed:
Fix Unix build with lunasvg

All that was needed was the --with-lunasvg

Randalphwa committed:
Fix path to lunasvg.h

Wrong path was used, and therefore autoconf created an invalid configure

Randalphwa committed:
Disable lunasvg by default and require C++17 when enabled

LunaSVG requires C++17 or later to compile, but was enabled by
default when wxUSE_ALL_FEATURES=yes, causing linker errors in builds
using C++11 or C++14.

Changes:
- Set DEFAULT_wxUSE_LUNASVG=no to disable by default
- Add C++ version checks that produce clear error messages when
  --with-lunasvg is used with --with-cxx=11 or --with-cxx=14
- Users must now explicitly request --with-lunasvg and use
  --with-cxx=17 or --with-cxx=20

This fixes build failures in CI where builds without explicit
--with-lunasvg were attempting to link against lunasvg symbols
that either failed to compile or had ABI incompatibilities.

Randalphwa committed:
Merge Maarten's lunasvg commits

Mostly these are a straight update from Maarten's fork, with a few like configure and configure.ac with changes on both sides.

Randalphwa committed:
Match other errors about required compiler version

Per PBfordev review suggestion.

Randalphwa committed:
Remove warning about compiler version

Now that the default value for wxUSE_LUNASVG is no, the various build systems should never be trying to build without a user changing it's value to "yes".

Randalphwa commented:
> I tried with configure on Windows and it doesn't work yet. `wxlunasvg` is not built and linked to, because it is not in the makefiles. You'll need to use `bakefile` to regenerate the makefiles after you modify `.bkl` files.
> 
> I did that myself and after configuring and running make it failed with `Makefile:2575: *** missing separator. Stop.`. There are some problems with the bakefile that results in a broken makefile. I think you used lexilla/scintilla as example for `lunasvg.bkl` because they way it is integrated now is not how the other third-party libraries are added. I changed it by using webp/png as example.
> 
> I committed my fixes to this branch. The first commit is my previously mentioned cmake change, then the bakefile/configure changes, and then a commit with all the generated files. The bakefile might need some more improving, but it works for me now! https://github.com/MaartenBent/wxWidgets/commits/lunasvg/

The problem with changing options.cmake to use "wx_add_thirdparty_library" is that it now defaults to builtin and _always_ attempts to build lunasvg on Windows, and therefore by default the build always fails because you can't build the library unless you set the compiler to C++17. That's why I didn't use it before. That's because of:

```
function(wx_add_thirdparty_library var_name lib_name help_str)
    cmake_parse_arguments(THIRDPARTY "" "DEFAULT;DEFAULT_APPLE;DEFAULT_WIN32" "" ${ARGN})

    if(NOT wxUSE_SYS_LIBS)
        set(thirdparty_lib_default builtin)
    elseif(THIRDPARTY_DEFAULT)
        set(thirdparty_lib_default ${THIRDPARTY_DEFAULT})
    elseif(THIRDPARTY_DEFAULT_APPLE AND APPLE)
        set(thirdparty_lib_default ${THIRDPARTY_DEFAULT_APPLE})
    elseif(THIRDPARTY_DEFAULT_WIN32 AND WIN32)
        set(thirdparty_lib_default ${THIRDPARTY_DEFAULT_WIN32})
    elseif(UNIX AND NOT APPLE)
        # Try sys libraries for MSYS and CYGWIN
        set(thirdparty_lib_default sys)
    elseif(WIN32 OR APPLE)
        # On Windows or apple platforms prefer using the builtin libraries
        set(thirdparty_lib_default builtin)
    else()
        set(thirdparty_lib_default sys)
    endif()

    wx_option(${var_name} ${help_str} ${thirdparty_lib_default}
        STRINGS builtin sys OFF)
```

The function lists OFF as an option, but cmake_parse_arguments and the conditionals below ignore it, so it defaults to builtin.

Randalphwa committed:
Support OFF as a valid option for thirdparty libs

Randalphwa committed:
Set wxUSE_LUNASVG to OFF

Valid option now that wx_add_thirdparty_library supprots it

Randalphwa committed:
Add check for invalid SVG data

vadz commented:
`wxBitmapBundle` is supposed to have a single invalid state, with `m_impl == nullptr`, so `IsOk()` doesn't need to and shouldn't be virtual. Instead, `wxBitmapBundle::FromSVG()` needs to check if creating the impl object was successful and return `wxBitmapBundle()` if it was not, like the existing `wxBitmapBundle::FromSVG()` implementation does.

Randalphwa committed:
Handle invalid SVG contents

Check if created impl was valid, return wxBitmapBundle() if not (see vadz review)

Removed bogus assertion check for file contents per vadz review

Randalphwa committed:
Update configure

Run autoconf version 2.69

Randalphwa committed:
Fix failing build when 'all' specified

The previous configuration was always building wxlunasvg when "all" was specified, even if the compiler didn't support it. This changes that so that only when both monolith and USE_LUNASVG are true will the wxlunasvg library be linked in.

Randalphwa committed:
bakefile regenerated files

Randalphwa committed:
Revert auto-format, reapply handling invalid SVG contents

Commit e748eeb743 mistakenly reformatted the entire file which was not only completely unintentional, it also broke CI builds. The commit removes all that reformatting and then reapplies the commits in c7bf332747.

Randalphwa committed:
Fix wxlunasvg linking in non-monolithic shared builds

The LIB_LUNASVG_IF_MONO variable was incorrectly used in the wx_append_nomono template, which is for non-monolithic builds where that variable is empty. Changed to use LIB_LUNASVG directly.

Randalphwa committed:
Move stdio.h header to the top of the file

This should pull in the necessary files to fix the isan() issue.

Randalphwa committed:
Implement workaround for GCC < 5

This adds a workaround for GCC < 5 that brings std::isnan into the global namespace with using std::isnan; right before nanosvg headers are included. This should fix the GCC 4.8 build without requiring a major file reorganization.  This avoids a major reorganization -- though if this doesn't work, I'll add that next.

vadz commented:
Thanks, AppVeyor failure is surely spurious, looks like something is seriously wrong with some infrastructure somewhere (launchpad.net doesn't work either).

I'd really like to get some feedback before merging this, but as I probably won't, I'll just merge it soon after one last superficial review.

MaartenBent commented:
I did test it again (on Windows, with CMake and configure), and there all seemed to work fine.

vadz commented:
I've tried to improve the docs a bit, but I think we need to merge this and see how LunaSVG development goes before addressing the other question, so I'll do it now.

Thanks again for all your work on this, @Randalphwa!
